File: Solutions/199-binary-tree-right-side-view/src.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

def bfs(root, ans):
    queue = []
    depth = []
    queue.append(root)
    depth.append(1)

    lastNodeDepth = 1
    lastTopNode = root

    while(len(queue) != 0):

        topNode = queue.pop(0)
        topNodeDepth = depth.pop(0)
        
        # print(topNode.val)
        # print(topNode.left)
        # print(topNode.right)
        
        if(topNode.left != None):
            queue.append(topNode.left)
            depth.append(topNodeDepth + 1)
        if(topNode.right != None):
            queue.append(topNode.right)
            depth.append(topNodeDepth + 1)

        # print(topNodeDepth)
        # print(lastNodeDepth)
        
        if(topNodeDepth != lastNodeDepth):
            ans.append(lastTopNode.val)
            
        lastTopNode = topNode
        lastNodeDepth = topNodeDepth
        
        if(len(queue) == 0):
            ans.append(topNode.val)

    return ans
    
class Solution:
    
    def rightSideView(self, root: TreeNode) -> List[int]:
        if(root == [] or root == None):
            return []
        ans = []
        bfs(root, ans)
        return ans

File: Solutions/5/src.cpp
/*
法一：暴力法
法二：动态规划
复杂度分析： 时间O(n^2) 空间O(n^2)  空间复杂度优化？
法三：中心扩展法
法四：Manacher（马拉车）算法*
*/

class Solution {
public:
    string longestPalindrome(string s) {
        bool P[1024][1024];
        string ans;
        int maxlen = 1;
        memset(P, false, sizeof(P) );
        if(s.size() != 0) ans.assign(s, 0, 1);
        for(int i = 0; i < s.size(); i++)
        {
            P[i][i] = true;
            if(i + 1 < s.size() && s[i] == s[i + 1] )
            {
                P[i][i + 1] = true;
                if(2 > maxlen)
                {
                    maxlen = 2;
                    ans.assign(s, i, 2);
                }
            }
        }
        for(int len = 3; len <= s.size(); len++)
        {
            for(int i = 0; i <= s.size() - len; i++)
            {
                int left = i, right = i + len - 1;
                if(P[left + 1][right - 1] == true && s[left] == s[right])
                {
                    P[left][right] = true;
                    if(len > maxlen)
                    {
                        maxlen = len;
                        ans.assign(s, i, maxlen);
                    }
                }
            }
        }
        return ans;
    }
};
